torchsynth/module.py

```diff
@@ -227,7 +227,65 @@ def update_device(self, device=None):
             parameter.parameter_range.to(device)
 
 
-class ADSR(SynthModule):
+class ControlRateModule(SynthModule):
+    """
+    Control rate module is an abstract base class for SynthModules that run
+    at a control rate as opposed to full audio rate.
+
+    Optionally, the output control signal can be upsampled to the correct
+    sample rate and buffer size.
+    """
+
+    def __init__(
+        self,
+        synthglobals: SynthGlobals,
+        upsample: Optional[bool] = False,
+        **kwargs: Dict[str, T],
+    ):
+        super().__init__(synthglobals, **kwargs)
+        self.upsample = upsample
+        if upsample:
+            self.upsample = torch.nn.Upsample(
+                self.synthglobals.buffer_size, mode="linear", align_corners=True
+            )
+
+    @property
+    def sample_rate(self) -> T:
+        """
+        The internal sample rate for this module is control rate
+        """
+        assert self.synthglobals.control_rate.ndim == 0
+        return self.synthglobals.control_rate
```

Comment:
Would like your guys opinion on this idea for overriding sample rate and buffer size properties to return the control rate and control buffer size. The benefit of this is that within the module the code will just run nicely at the correct rate and any calls to helper functions like seconds_to_samples will work without having to think about the different rates. The downside is that it is maybe a bit too cute and could lead to confusion, for example if you run this module with upsampling then from an outside perspective the sample_rate property won't seem right. But internally it would be.

Reply:
Hmm how hard would be only to have sample_rate for SignalModules and control_rate for control_modules?

I would assume sample_rate of a ControlRateModule would including the upsampling

Reply:
Not hard. ControlRateModule derive from SynthModule currently and has access to everything.